firmware_391/control.cpp

```diff
@@ -0,0 +1,280 @@
+#include "control.h"
+#include <time.h> 
+
+Switches             switches(0xFF200000);
+RS232               bluetooth(0xFF200080);
+// RS232                    wifi(0xFF200088);
+WaveformPlayer waveformplayer(0xFF200090);
+SDRAM                   sdram(0xC0000000);
+Parser                  parser(&bluetooth);
+
+control::control(){
+    this->playing = false;
+    this->reset_bt_parser();
+}
+
+
+int control::commence(){
+    char* data;
+    unsigned int data_len;
+
+    int parse_result;
+
+    while(true){
+        data_len = parser.getdata(&data); // start every 
+        if (data_len > 0){
+            for(unsigned int i = 0; i < data_len; i++){
+                parse_result = this->parse_bluetooth(data[i]);
+                if (parse_result < 0) continue;
+                this->execute_cmd(this->last_command);
+            }
+        }
+        else {
+            if (this->playing){
+                // TODO: check for stop command in switches
+                continue;
+            }
+            else {
+                // TODO: maybe add switch input?
+                continue;
+            }
+        }
+    }
+}
+
+void control::reset_bt_parser(){
+    this->parse_buf_i = 0;
+    this->state = PREF;
+    this->in_progress.param1 = 0.0;
+    this->in_progress.param2 = 0.0;
+    this->in_progress.param3 = 0.0;
+}
+
+int control::parse_bluetooth(char c){
```

Comment:
Please add the control code to place the values we receive into an array of type `waveform_element` and a depth of... idk 16?

Thank you

Reply:
Something like this 

```
    struct waveform_element el = {.type = sine, .periodic = {.freq = 40, .amplitude = 1, .offset = 90}};
    waveforms[i++] = el;
```